skbeam/core/accumulators/arrayProjector.py

```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env python
+
+import math
+import numpy as np
+
+"""
+authors: Mikhail Dubrovin, TJ Lane, Christopher O'Grady
+"""
+
+class ArrayProjector(object):
+
+    def __init__(self, bin_values, weights=None, nbins=101, norm=True):
+        """
+        Parameters
+        ----------
+        bin_values : np.ndarray
+            For each pixel, this is the value of that determines which bin
+            a pixel's intensity will contribute to
+        weights : np.ndarray
+            A weight for each pixel with the same shape as bin_values (can be zero to ignore a pixel)
+        nbins : int
+            The number of bins to employ. If `None` guesses a good value.
+        """
+
+        self.bin_values = bin_values
+        self.weights = weights
+        self.nbins = nbins
+        self.norm = norm
+
+        self.binvalRange = self.bin_values.max() - self.bin_values.min()
+        self.bin_width = self.binvalRange / (float(nbins) - 1)
+
+        self._bin_assignments = np.floor( (self.bin_values - self.bin_values.min()) / self.bin_width ).astype(np.int32)
+        if weights is None:
+            self._normalization_array = (np.bincount( self._bin_assignments.flatten() ) \
+                                             + 1e-100).astype(np.float)
+        else:
+            self._normalization_array = (np.bincount( self._bin_assignments.flatten(), weights=self.weights.flatten() ) \
+                                             + 1e-100).astype(np.float)
+
+        assert self.nbins >= self._bin_assignments.max() + 1, 'incorrect bin assignments'
+        self._normalization_array = self._normalization_array[:self.nbins]
+
+        return
+
+    def __call__(self, image):
+        """
+        Bin pixel intensities.
+        
+        Parameters
+        ----------            
+        image : np.ndarray
+            The intensity at each pixel
+        Returns
+        -------
+        bin_values : np.ndarray
+            The average intensity in the bin
+        """
+
+        if not (image.shape == self.bin_values.shape):
+            raise ValueError('`image` and `bin_values` must have the same shape',image.shape,self.bin_values.shape)
+
+        if self.weights is None:
+            weights = image.flatten()
+        else:
+            weights = image.flatten() * self.weights.flatten()
+            if not (image.shape == self.weights.shape):
+                raise ValueError('`image` and `weights` must have the same shape')
+
+        bin_values = np.bincount(self._bin_assignments.flatten(), weights=weights)
+
+        if self.norm:
+            bin_values /= self._normalization_array
```

Comment:
Add `from __future__ import division` so that this will *always* be float division